psydac/ddm/cart.py

```diff
@@ -696,20 +693,38 @@ def topetsc( self ):
     #---------------------------------------------------------------------------
     def coords_exist( self, coords ):
 
-        return all( P or (0 <= c < d) for P,c,d in zip( self._periods, coords, self._dims ) )
+        return all( P or (0 <= c < d) for P,c,d in zip( self._periods, coords, self._nprocs ) )
 
     #---------------------------------------------------------------------------
     def get_shift_info( self, direction, disp ):
 
+        if len(self._shift_info) == 0:
+            for axis in range( self._ndims ):
+                for d in [-1,1]:
+                    self._shift_info[ axis, d ] = \
+                            self._compute_shift_info( axis, d )
+
         return self._shift_info[ direction, disp ]
 
+    #---------------------------------------------------------------------------
+    def get_shift_info_non_blocking( self, shift ):
+
+        if len(self._shift_info_non_blocking) == 0:
+            zero_shift = tuple( [0]*self._ndims )
+            for sh in product( [-1,0,1], repeat=self._ndims ):
+                if sh == zero_shift:
+                    continue
+                self._shift_info_non_blocking[sh] = self._compute_shift_info_non_blocking( sh )
+
+        return self._shift_info_non_blocking[ shift ]
```

Comment:
because we have two different types of communications, depending on what the user wants in the future we might add others, so it makes sense to compute the ones that we need only, and here also we do it once and for all.
as for the computations, yes it slows things up a little bit in the github actions